packages/contracts-bedrock/src/L2/GasTank.sol

```diff
@@ -0,0 +1,250 @@
+
+        emit WithdrawalFinalized(msg.sender, _to, amount);
+    }
+
+    /// @notice Authorizes a message to be claimed by the relayer
+    /// @param _messageHash The hash of the message to authorize
+    function authorizeClaim(bytes32 _messageHash) external {
+        authorizedMessages[msg.sender][_messageHash] = true;
+
+        emit AuthorizedClaim(msg.sender, _messageHash);
+    }
+
+    /// @notice Relays a message to the destination chain
+    /// @param _id The identifier of the message
+    /// @param _sentMessage The sent message event payload
+    function relayMessage(
+        Identifier calldata _id,
+        bytes calldata _sentMessage
+    )
+        external
+        returns (uint256 gasCost_, bytes32[] memory nestedMessageHashes_)
+    {
+        uint256 initialGas = gasleft();
+
+        bytes32 messageHash = _getMessageHash(_id.chainId, _sentMessage);
+
+        uint240 nonceBefore = _getMessengerNonce();
+
+        MESSENGER.relayMessage(_id, _sentMessage);
+
+        // Get the amount of nested messages by getting the nonce increment
+        uint256 nonceDelta = _getMessengerNonce() - nonceBefore;
+
+        nestedMessageHashes_ = new bytes32[](nonceDelta);
+
+        for (uint256 i; i < nonceDelta; i++) {
+            nestedMessageHashes_[i] = MESSENGER.sentMessages(nonceBefore + (i + 1));
+        }
+
+        // Get the gas used
+        gasCost_ = _cost(initialGas - gasleft(), block.basefee) + _relayOverhead(nestedMessageHashes_.length);
+
+        // Emit the event with the relationship between the origin message and the destination messages
+        emit RelayedMessageGasReceipt(messageHash, msg.sender, gasCost_, nestedMessageHashes_);
+    }
+
+    /// @notice Claims repayment for a relayed message
+    /// @param _id The identifier of the message
+    /// @param _gasProvider The address of the gas provider
+    /// @param _payload The payload of the message
+    function claim(Identifier calldata _id, address _gasProvider, bytes calldata _payload) external {
+        // Ensure the origin is a gas tank deployed with the same address on the destination chain
+        if (_id.origin != address(this)) revert InvalidOrigin();
+
+        // Validate the message
+        ICrossL2Inbox(Predeploys.CROSS_L2_INBOX).validateMessage(_id, keccak256(_payload));
+
+        (bytes32 originMessageHash, address relayer, uint256 relayCost, bytes32[] memory destinationMessageHashes) =
+            decodeGasReceiptPayload(_payload);
+
+        if (!authorizedMessages[_gasProvider][originMessageHash]) revert MessageNotAuthorized();
+
+        if (claimed[originMessageHash]) revert AlreadyClaimed();
+
+        uint256 destinationMessageHashesLength = destinationMessageHashes.length;
+
+        // Authorize nested messages by the same gas provider
+        for (uint256 i; i < destinationMessageHashesLength; i++) {
+            authorizedMessages[_gasProvider][destinationMessageHashes[i]] = true;
```

Comment:
@hexshire can you emit an `AuthorizedClaim` event for each of these?

Reply:
To add more context here, an event is needed here in order for the relayer to keep track of authorized claims for nested messages